### DISCUSSION TITLE
How to suppress "Expected component to be defined"?

### DISCUSSION THREAD

**Discussion by hahnbeelee** [mdx-js/mdx#2402] [Q&A]
My goal is to convert an MDX string to a React string. By following [this thread](https://github.com/orgs/mdx-js/discussions/1985), I have created this function

```typescript
import { evaluateSync } from '@mdx-js/mdx';
import { createElement } from 'react';
import { renderToString } from 'react-dom/server';
import * as runtime from 'react/jsx-runtime';

import { coreRemarkMdPlugins } from './remark.js';

export function mdxToReact(body: string) {
  const mdx = evaluateSync(body, {
    Fragment: undefined,
    jsx: undefined,
    jsxs: undefined,
    ...(runtime as any),
    remarkPlugins: coreRemarkMdPlugins,
  }).default;

  return renderToString(createElement(mdx));
}
```

Unfortunately when I put in a string like `"<RandomComponent>child</RandomComponent>"` it throws an error `Error: Expected component `RandomComponent` to be defined: you likely forgot to import, pass, or provide it.`.

All I want is to go from an mdx string to a react string. Is there anyway I can suppress this error? I read the options [here](https://mdxjs.com/packages/mdx/#processoroptions) but I wasn't able to draw any conclusions.

Thank you!

**Comment by wooorm** [comment]
`body` is already a string.
Why do you `evaluate`?
what do you mean “react string”? `<RandomComponent>child</RandomComponent>` is a react string?

**Reply by wooorm** [reply]
Maybe you are looking for this guide? https://mdxjs.com/guides/mdx-on-demand/

If not: I don’t really know what you want. Can you add some more info on what you want to do?

**Reply by hahnbeelee** [reply]
`compile` returns js. If I input `# Hello, world!` `compile` returns:
```
import {jsx as _jsx} from "react/jsx-runtime";
function _createMdxContent(props) {
  const _components = {
    h1: "h1",
    ...props.components
  };
  return _jsx(_components.h1, {
    children: "Hello, world!"
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
```
I want `<h1>Hello, world!</h1>`

To describe the "React string" that I want. I want it to process my MDX s.t. it's human readable for a frontend engineer to copy paste into their VS Code and it'll be a valid React component.

**Reply by hahnbeelee** [reply]
My implementation with `evaluateSync` does what I want it to do up to some extent. For example:
```typescript
it('should transform code', () => {
    expect(mdxToReact('`code`')).toBe('<p><code>code</code></p>');
  });
```
The test case above passes.
```typescript
it('should ignore components', () => {
    expect(mdxToReact('<RandomComponent>child</RandomComponent>')).toBe(
      '\n<p><RandomComponent>child</RandomComponent></p>\n'
    );
  });
```
But this one does not. This one throws "Error: Expected component `RandomComponent` to be defined: you likely forgot to import, pass, or provide it". Is there anyway I can get the above test case to pass?

**Reply by ChristianMurphy** [reply]
> Is there anyway I can get the above test case to pass?

The most direct way would be to pass a definition of `RandomComponent` into the `default`/render function (that comes from `evaluateSync`) as the `components` prop.

```ts
import { evaluateSync } from '@mdx-js/mdx';
import { createElement } from 'react';
import { renderToString } from 'react-dom/server';
import * as runtime from 'react/jsx-runtime';

import { coreRemarkMdPlugins } from './remark.js';

export function mdxToReact(body: string, components) {
  const mdx = evaluateSync(body, {
    Fragment: undefined,
    jsx: undefined,
    jsxs: undefined,
    ...(runtime as any),
    remarkPlugins: coreRemarkMdPlugins,
  }).default;

  return renderToString(createElement(mdx, { components }));
}
```

You could also use a `Proxy` to try to fill all components with some default value https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy
Though use that with caution, suppressing errors usually isn't a good approach.
Tests are supposed to be a reflection of how your users will interact with your system, the system silently crashing and not showing what the user wants _probably_ isn't a good reflection of that.

**Reply by hahnbeelee** [reply]
> the system silently crashing and not showing what the user wants probably isn't a good reflection of that.

I 100% agree. This is an intermediate step and there is logic later down the road that will ensure that the user is alerted about the crash.

> You could also use a Proxy to try to fill all components with some default value https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy

Could you give me some more pointers with this approach?

**Reply by wooorm** [reply]
> This is an intermediate step and there is logic later down the road that will ensure that the user is alerted about the crash.

> ```
> expect(mdxToReact('<RandomComponent>child</RandomComponent>')).toBe(
>   '\n<p><RandomComponent>child</RandomComponent></p>\n'
> );

It sounds like you’re trying to do something entirely different.
`evaluate` evaluates code. It doesn’t turn MDX syntax into React syntax.
Perhaps you can share more about what you actually want to do?

**Reply by hahnbeelee** [reply]
> Perhaps you can share more about what you actually want to do?

Sure, my end goal is to transform the following code so that it goes from this

```mdx
export const UserDefinedMdxContent = () => <Note>
  ### markdown
  _hello_
</Note>;

<UserDefinedMdxContent />
```

to this:
```mdx
export const UserDefinedMdxContent = () => <Note>
  <h3>markdown</h3>
  <b>hello</b>
</Note>;

<UserDefinedMdxContent />
```

because the first code block renders like so (not what I want):
<img width="788" alt="image" src="https://github.com/mdx-js/mdx/assets/55263191/53a044df-1d0a-4d24-803f-24ca7fa076d7">

but the second one renders like this (which is what I want):
<img width="785" alt="image" src="https://github.com/mdx-js/mdx/assets/55263191/c2b7e12d-c302-40a5-a6e7-a5b4886357d6">

Initially I thought I would be able to achieve this simply by using a markdown to html convertor like so:
```typescript
import { toHtml } from 'hast-util-to-html';
import { fromMarkdown } from 'mdast-util-from-markdown';
import { toHast } from 'mdast-util-to-hast';

export const mdToHtml = (markdown: string) => {
  const mdast = fromMarkdown(markdown);
  const hast = toHast(mdast);
  if (hast) {
    const html = toHtml(hast);
    return html;
  }
  throw Error('Could not parse from markdown to html');
};
```
but this also didn't yield what I wanted because this was my output:
```
console.log(mdToHtml('<RandomComponent>child</RandomComponent>'))
<p>child</p>
```
When I wanted to preserve my jsx component
```jsx
<p><RandomComponent>child</RandomComponent></p>
```
So when I say the following
> This is an intermediate step and there is logic later down the road that will ensure that the user is alerted about the crash.
This is how I see my steps:
1. Go through content and see if there are any exported components that are used.
2. If there are run the contents of the export through `mdxToHtml` so that the markdown is transformed into html so that it's displayed properly.
3. When the page is finally visited, the contents of my file will have all exports tranformed so that all markdown is changed to html. At this stage that transformed content will be put into a parser again. If any of the components are not defined at _this stage_ then the parsing will fail and it'll let the user know that any components that they are trying to use is unavailable. 

To the user, they won't have known that the content has gone through two passes and they'll be alerted normally. This is why I'm trying to ignore the "Expected component to be defined" error. But as you guys have said please let me know if there is another, better way for me to approach this other than what I've already attempted:
1. markdown to html using `hast-util-to-html`, `mdast-util-from-markdown`, `mdast-util-to-hast` (this approach removed my custom jsx components)
2. `evaluateSync` from `@mdx-js/mdx` + `createElement` and `renderToString` from `react` (this approach throws at my jsx components)

**Reply by wooorm** [reply]
> so that it goes from this

Why do you want people to write like that?
Why not tell people to write the second?
Or tell them to put it in a different file?
You can import MDX files in other MDX files.

**Reply by hahnbeelee** [reply]
> Why do you want people to write like that?
> Why not tell people to write the second?

Because of the beauty of mdx - it's a better user experience to use markdown over html tags. Telling our users to switch between using markdown and html syntax seems like a subpar user experience.

> Or tell them to put it in a different file?
> You can import MDX files in other MDX files.

I'm running a multi-tenant next.js application that is dependent on getStaticProps/next-mdx-remote. So I don't have access to a bundler or file system at the moment. So my goal right now is basically to recreate exports/imports from mdx files to other mdx files

**Reply by hahnbeelee** [reply]
It's why I asked [this](https://github.com/orgs/mdx-js/discussions/2401) and what I'm trying to achieve in this thread is this step of the import:

> Though the component will have translated `#` into an `<h1>` by this point.

**Reply by hahnbeelee** [reply]
Maybe a more concise way of asking my initial question is - is there a way to change all my markdown to html inside my mdx files without impacting (removing or throwing on) my jsx components?

**Reply by ChristianMurphy** [reply]
> is there a way to change all my markdown to html inside my mdx files without impacting (removing or throwing on) my jsx components?

Short answer is: no.
That is not what MDX is designed to do.

You could get close-ish, by adding a `rehype` plugin to snapshot the `hast` content. https://mdxjs.com/packages/mdx/
At this stage markdown has been translated to HTML, but custom elements have not been rendered yet (example https://mdxjs.com/playground/)

---

Though taking a step back, this is very much feel like an [XY Problem](https://xyproblem.info/).
Why test an output which no user will ever see? What property of your rendering pipeline are you trying to test here?

**Reply by wooorm** [reply]
> So I don't have access to a bundler or file system at the moment. 

It might be better to solve this problem 👆

> Telling our users to switch between using markdown and html syntax seems like a subpar user experience.

Important note: that isn’t HTML. It’s JSX. It’s a different language.

> Is there a way to change all my markdown to html inside my mdx files without impacting (removing or throwing on) my jsx components?

What is markdown?

```js
export const a = {'is *this* markdown?': '*this*?'}
```

```js
export const b = 'is *this* markdown?'
```

```js
export const c = is *this* markdown?
```

```js
export const d = <>is *this* markdown?</>
```

```js
export const e = <>{'this?'}</>
```

```js
{'*this*?'}
```

```js
{'# this?'}
```

I would strongly recommend not inventing a new language, that doesn’t work in other places.
I do recommend solving the file system.

*If* you *must* do this, I recommend a tagged template:

```mdx
export function SomeComponent() {
  return mdx`# hi`
}
```

Something along the lines of a plugin looking for `TaggedTemplateExpression`s with the tag `mdx` then use `unified().use(remarkParse).use(remarkMdx)` to get the tree and stitch the ASTs together.
